File: backend-java/src/main/java/com/github/pascualex/oc/model/Course.java
package com.github.pascualex.oc.model;

public class Course {

    private final int id;
    private final String title;
    private final CourseLevel level;
    private final int hours;
    private final boolean active;
    private final Teacher teacher;

    public Course(int id, String title, CourseLevel level, int hours, boolean active, Teacher teacher) {
        this.id = id;
        this.title = title;
        this.level = level;
        this.hours = hours;
        this.active = active;
        this.teacher = teacher;
    }

    public Course(int id, String title, String level, int hours, boolean active, Teacher teacher) {
        this(id, title, CourseLevel.valueOf(level), hours, active, teacher);
    }

    public Course(String title, CourseLevel level, int hours, boolean active, Teacher teacher) {
        this(-1, title, level, hours, active, teacher);
    }

    public Course() {
        this(-1, "", CourseLevel.UNKNOWN, -1, false, null);
    }

    public int getId() {
        return id;
    }

    public String getTitle() {
        return title;
    }

    public CourseLevel getLevel() {
        return level;
    }

    public int getHours() {
        return hours;
    }

    public boolean isActive() {
        return active;
    }

    public Teacher getTeacher() {
        return teacher;
    }
}

File: backend-node/src/model/Course.ts
import { CourseLevel, parseCourseLevel } from './CourseLevel';
import { Teacher, bodyToTeacher } from './Teacher';

export class Course {
  constructor(
    public readonly id: number,
    public readonly title: string,
    public readonly level: CourseLevel,
    public readonly hours: number,
    public readonly active: boolean,
    public readonly teacher: Teacher,
  ) { }
}

export interface CourseRow {
  id: number,
  title: string,
  level: string,
  hours: number,
  active: number,
  teacher_id: number,
  teacher_name: string
}

export function bodyToCourse(body: any): Course {
  if (typeof body.title !== 'string' || body.title.length < 3) return null;
  const level: CourseLevel = parseCourseLevel(body.level);
  if (level === CourseLevel.Unknown) return null;
  if (typeof body.hours !== 'number' || body.hours <= 0) return null;
  if (typeof body.active !== 'boolean') return null;
  const teacher: Teacher = bodyToTeacher(body.teacher);
  if (!teacher) return null;
  if (typeof body.id === 'number' && body.id > 0) {
    return new Course(body.id, body.title, level, body.hours, body.active, teacher);
  } else {
    return new Course(undefined, body.title, level, body.hours, body.active, teacher);
  }
}

export function rowToCourse(row: CourseRow): Course {
  const level: CourseLevel = parseCourseLevel(row.level);
  const active: boolean = row.active !== 0;
  const teacher: Teacher = new Teacher(row.teacher_id, row.teacher_name, undefined);
  return new Course(row.id, row.title, level, row.hours, active, teacher);
}